byron/semantics/executable-spec/src/Control/State/Transition/Trace/Generator/QuickCheck.hs

```diff
@@ -188,22 +189,44 @@ forAllTrace
 forAllTrace baseEnv maxTraceLength traceGenEnv =
   forAllTraceFromInitState baseEnv maxTraceLength traceGenEnv Nothing
 
--- | See 'Test.QuickCheck.shrink'.
+-- | Shrink a trace by shrinking the list of signals and reconstructing traces from these
+-- shrunk lists of signals. 
+--
+-- When shrinking a trace that is failing some property (often stated in terms of a signal in the trace)
+-- then the most recent signal is likely crucial to the failure of the property and must be preserved
+-- in the shrunk traces.
 shrinkTrace
   :: forall sts traceGenEnv
    . (HasTrace sts traceGenEnv
      )
   => BaseEnv sts
   -> Trace sts
   -> [Trace sts]
-shrinkTrace baseEnv tr = interpretSTS @sts @traceGenEnv baseEnv
-    $ Trace.closure env st0 `traverse` traceSignalsShrinks
+shrinkTrace baseEnv tr =
+    interpretSTS @sts @traceGenEnv baseEnv
+    $ Trace.closure env st0 `traverse` shrunk
   where
     env = Trace._traceEnv tr
     st0 = Trace._traceInitState tr
-    signals = Trace.traceSignals Trace.OldestFirst tr
-    traceSignalsShrinks =
-      QuickCheck.shrinkList (shrinkSignal @sts @traceGenEnv) signals
+    signals = Trace.traceSignals Trace.NewestFirst tr
+
+    shrunk = sortOn length (shrinkSignals signals)
+
+    -- Shrink a list of signals such that we preserve the most recent signal in the shrunk lists.
+    -- This shrinker
+    --   - recursively omits all but the most recent signal
+    --   - builds up lists of signals starting with the most recent signal and
+    --     building up a list excluding the first (oldest) signal
+    shrinkSignals (sn:_last:[]) =
+      [[sn]]
+    shrinkSignals (sn:sm:sigs) =
+      [[sn]] -- a trace with only the most recent signal
```

Comment:
Why will this provide a meaningful shrink? If we take the newest signal (the last signal we apply) and apply it to `s0` that signal will probably need state information that won't be in `s0`. So it seems `closure` would fail and we will be performing unnecessary work.

Reply:
yes the shrink to [sn] has a low likelihood to succeed but has a high pay-off, so I think it's worth trying. (also `closure` just ignores signals that fail on some STS predicate, so it's ok on that front)